File: algorithms/code/leetcode/lc007_reverse_integer/lc007_reverse_integer.py
#!/usr/bin/env python3
# https://leetcode.com/problems/reverse-integer/

import unittest


class Solution:
    def reverse(self, x: int) -> int:
        if x == -2147483648:
            return 0
        negative = x < 0
        if negative:
            x = -x
        digits = []
        while x > 0:
            digits.append(x % 10)
            x //= 10
        limits = [2, 1, 4, 7, 4, 8, 3, 6, 4, 7]
        length = len(limits)
        padding = length - len(digits)
        prefix = padding * [0]
        digits = prefix + digits
        for i in range(length):
            if digits[i] > limits[i]:
                return 0
            if digits[i] < limits[i]:
                break
        result = 0
        for digit in digits:
            result = result * 10 + digit
        if negative:
            result = -result
        return result


class TestCode(unittest.TestCase):
    def test_minus2000000002(self):
        self.assertEqual(Solution().reverse(2000000002), 2000000002)

    def test_minus2147483648(self):
        self.assertEqual(Solution().reverse(-2147483648), 0)

    def test_1000000003(self):
        self.assertEqual(Solution().reverse(1000000003), 0)

    def test_1534236469(self):
        self.assertEqual(Solution().reverse(1534236469), 0)

    def test_minus321(self):
        self.assertEqual(Solution().reverse(-321), -123)

File: algorithms/code/leetcode/lc007_reverse_integer/lc007_reverse_integer.rs
// https://leetcode.com/problems/reverse-integer/

pub struct Solution {}

impl Solution {
    pub fn reverse(mut x: i32) -> i32 {
        if x == i32::MIN {
            return 0;
        }
        let negative = x < 0;
        if negative {
            x = -x;
        }
        let mut digits: Vec<i32> = vec![];
        while x > 0 {
            digits.push(x % 10);
            x /= 10;
        }
        let limits = [2, 1, 4, 7, 4, 8, 3, 6, 4, 7];
        let length = limits.len();
        let padding = length - digits.len();
        let mut prefix: Vec<i32> = vec![0; padding];
        prefix.extend(digits);
        digits = prefix;
        for i in 0..length {
            if digits[i] > limits[i] {
                return 0;
            }
            if digits[i] < limits[i] {
                break;
            }
        }
        let mut result: i32 = 0;
        for digit in digits.iter() {
            result = result * 10 + *digit;
        }
        if negative {
            result = -result;
        }
        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_minus2000000002() {
        assert_eq!(Solution::reverse(2000000002), 2000000002);
    }

    #[test]
    fn test_minus2147483648() {
        assert_eq!(Solution::reverse(-2147483648), 0);
    }

    #[test]
    fn test_1000000003() {
        assert_eq!(Solution::reverse(1000000003), 0);
    }

    #[test]
    fn test_1534236469() {
        assert_eq!(Solution::reverse(1534236469), 0);
    }

    #[test]
    fn test_minus321() {
        assert_eq!(Solution::reverse(-321), -123);
    }
}
